### PR TITLE
(#663) sovereign cloud support

### PR DESCRIPTION
**Pull request by adrianhall**

### CODE REVIEW

File: _includes/tables/environment_variables.md [hunk 1]
```diff
@@ -19,5 +19,6 @@
 | AZURE_TRACING_DISABLED        | Disables tracing                       |
 | **General SDK Configuration**                                         ||
 | AZURE_CLOUD                   | Name of the sovereign cloud            |
+| ARM_DATA_ENDPOINT             | URI to load the list of endpoints      |
```
Comment (JonathanGiles) [inline _includes/tables/environment_variables.md]: I know 'Azure' is the first part of ARM, but I wonder if we should be starting our environment variables with AZURE for consistency with others, and to make it clearer to people what the environment variables pertains to (assuming this environment variables isn't a pre-existing one used elsewhere)?

Comment (adrianhall) [reply _includes/tables/environment_variables.md]: It's a specific environment variable shared with other tooling, so it needs to be identical.  I will validate the name prior to merge.

File: docs/general/azurecore.md [hunk 2]
```diff
@@ -261,6 +261,43 @@ Global configuration refers to configuration settings that are applied to all ap
...
 
+### Sovereign cloud and secure endpoints
+
+The Azure Cloud is not just one cloud.  There are sovereign clouds (such as those in China and Germany), government clouds, and Azure Stack - an on-premise implementation of Azure.  There are three primary differences between sovereign clouds and the Azure public clouds:
+
+1. Not all services may be available in sovereign clouds.
+2. The list of supported API versions may be different.
+3. The endpoints used will be different.
+
+The majority of services are configured using an endpoint definition (which is generally a fully-qualified URI).  However, some services have implicit endpoints defined (for example, Azure Identity uses `https://login.microsoftonline.com`) and some services construct an endpoint based on a friendly name.  There are approximately 15-20 services for which one of these cases hold.
```
Comment (johanste) [inline docs/general/azurecore.md]: Should we specify which cloud Azure Identity uses the given endpoint for? E.g.

> Azure Identity uses `https://login.microsoftonline.com` for public Azure.

File: docs/general/azurecore.md [hunk 3]
```diff
@@ -261,6 +261,43 @@ Global configuration refers to configuration settings that are applied to all ap
...
+|`AzureCloud`|Default cloud instance. Used unless overridden by application.|
+|`AzureChinaCloud`|https://azure.microsoft.com/en-us/global-infrastructure/china/|
+|`AzureUSGovernment`|https://azure.microsoft.com/en-us/global-infrastructure/government/|
+|`AzureGermanCloud`|https://azure.microsoft.com/en-us/global-infrastructure/germany/|
+
+In the latter case, the developer will set the well-known URI in the `ARM_DATA_ENDPOINT_URL` setting.  The application will then download a JSON file that identifies the endpoints for each service.
+
+In terms of precedence, use the following:
+
+1. Developer-provided endpoint information.
```
Comment (johanste) [inline docs/general/azurecore.md]: All of the below are "Developer-provided" endpoint information, no? I believe that what we are trying to say is that values provided explicitly in API calls overrides other forms of configuration. I'm pretty sure we have an exact phrasing we can use somewhere in the environments section of the guidance.

File: docs/general/azurecore.md [hunk 4]
```diff
@@ -261,6 +261,43 @@ Global configuration refers to configuration settings that are applied to all ap
 
 {% include requirement/MUST id="azurecore-config-opt-out" %} provide a method of opting out from importing system settings and environment variables into the configuration.
 
+### Sovereign cloud and secure endpoints
```
Comment (johanste) [inline docs/general/azurecore.md]: What does "secure endpoints" refer to?

File: docs/general/design.md [hunk 5]
```diff
@@ -81,6 +81,49 @@ The following are standard verb prefixes.  You should have a good (articulated)
 
 {% include requirement/MUST id="general-client-feature-support" %} support 100% of the features provided by the Azure service the client library represents. Gaps in functionality cause confusion and frustration among developers.
 
+## Sovereign clouds and default endpoints
+
+The Azure Cloud is not just one cloud.  There are sovereign clouds (such as those in China and Germany), government clouds, and Azure Stack - an on-premise implementation of Azure.  There are three primary differences between sovereign clouds and the Azure public clouds:
```
Comment (johanste) [inline docs/general/design.md]: Is it intentional that we are repeating this section?

File: docs/general/azurecore.md (hunk 4 above)
Comment (adrianhall) [reply docs/general/azurecore.md]: Fair point.  It was copied from your terminology.  Removed.

File: docs/general/azurecore.md (hunk 3 above)
Comment (adrianhall) [reply docs/general/azurecore.md]: I've adjusted - see what you think.

File: docs/general/design.md (hunk 5 above)
Comment (adrianhall) [reply docs/general/design.md]: Yes - it's intentional.  A lot of what we do is in Azure Core, but not all.  This section is for library designers.  The azurecore section is for azure core developers.

**Comment by KrzysztofCwalina** [comment]
The guidelines look good, but it seems like a lot of work to support and test it all. Do we have this work planned? I am hesitant to have "MUST" guidelines that we ourselves don't have time/resources to follow.

**Pull request merged**